src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala

```diff
@@ -43,16 +82,43 @@ class AkkaHttpSecurity[P <: CommonProfile](config: Config) {
    * this does not apply any authorization ofr filtering.
    */
   def withAuthentication(
-    clients: String = null /* Default null, meaning all defined clients */ ,
-    multiProfile: Boolean = true): Directive1[AuthenticatedRequest] =
+               clients: String = null /* Default null, meaning all defined clients */ ,
+               multiProfile: Boolean = true,
+               enforceFormEncoding: Boolean = false, //Force form parameters to be passed for authentication or the request fails
+             ): Directive1[AuthenticatedRequest] =
     new Directive1[AuthenticatedRequest] {
       override def tapply(innerRoute: Tuple1[AuthenticatedRequest] => Route): Route = { ctx =>
-        val context = AkkaHttpWebContext(ctx.request)
-        securityLogic.perform(context, config, (context: AkkaHttpWebContext, profiles: util.Collection[CommonProfile], parameters: AnyRef) => {
-          val authenticatedRequest = AuthenticatedRequest(context, profiles.asScala.toList)
-          innerRoute(Tuple1(authenticatedRequest))(ctx)
-          // TODO: Add response headers from the AkkaHttpWebContext and put them into the response
-        }, actionAdapter, clients, "", "", multiProfile)
+        import ctx.materializer
+
+        /**
+          * First try to extract authentication credentials from form parameters using Akka's StrictForm unmarshallers.
+          * If that fails, either form encoding is enforced, in which case the request fails, or the request proceeds.
+          * If the request proceeds, other ways (e.g. basic auth) are assumed to be configured in pac4j in order to pass
+          * credentials.
+          */
+        Unmarshal(ctx.request.entity).to[StrictForm].fast.flatMap { form =>
+          val fields = form.fields.collect {
+            case (name, field) if name.nonEmpty =>
+              Unmarshal(field).to[String].map(fieldString => (name, fieldString))
+          }
+          Future.sequence(fields)
+        }.map { fields =>
+          AkkaHttpWebContext(ctx.request, fields)
+        }.recoverWith { case e =>
+          if(enforceFormEncoding) {
```

Comment:
This recovers all errors happening during AkkaHttpWebContext construction, i'd move the recovery before the AkkaHttpWebContextConstruction. You can also turn the whole form field reading into a 'smart constructor' of AkkaHttpSecurity (one that takes just the HttpRequest and the enforceFormEncoding and returns a Future[AkkaHttpRequestContext])